net: make m_nodes_mutex non-recursive

vasild opened the pull request:
The only case of a recursive lock was a nested `ForNode()` call to trim
the size of `lNodesAnnouncingHeaderAndIDs` to `<= 3`. This need not be
nested, so take it out.

Before:
```
fornode(newnode)
    if (size >= 3)
        fornode(front) handle removal of front
        pop front
    push back newnode
```

After:
```
fornode(newnode)
    push back newnode
if (size > 3)
    fornode(front) handle removal of front
    pop front
```

`lNodesAnnouncingHeaderAndIDs` is protected by `cs_main` which is locked
during the entire operation.

Partially resolves: https://github.com/bitcoin/bitcoin/issues/19303

---

_This PR included https://github.com/bitcoin/bitcoin/pull/32326 (first 3 commits in this PR). That PR was merged first, so the size of this was reduced._

Bonobirho99 reviewed (commented):
.

hebasto commented:
> Partially resolves: #19303

Concept ACK on that.

vasild commented:
`98bba932bf...22d4c57a99`: rebase due to conflicts

hebasto commented:
Time to rebase once more?

vasild commented:
`aaa75c1a41...492be23a18`: rebase due to conflicts

hodlinator reviewed (commented):
Concept ACK 492be23a1873d7165e6cc75d1282f4feb63f87b9

vasild committed:
net: drop the only recursive usage of CConnman::m_nodes_mutex

The only recursive usage of `CConnman::m_nodes_mutex` is from
`PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs()` which uses
nested calls to `CConnman::ForNode()` to trim the size of
`lNodesAnnouncingHeaderAndIDs` to `<= 3`. This need not be nested, so
take it out.

Before:
```
fornode(newnode)
    if (size >= 3)
        fornode(front) handle removal of front
        pop front
    push back newnode
```

After:
```
fornode(newnode)
    push back newnode
if (size > 3)
    fornode(front) handle removal of front
    pop front
```

`lNodesAnnouncingHeaderAndIDs` is protected by `cs_main` which is locked
during the entire operation.

vasild committed:
net: make CConnman::m_nodes_mutex non-recursive

This change includes `s/RecursiveMutex/Mutex/` and a pile of
annotations to keep the compiler happy after the type change.

Partially resolves: https://github.com/bitcoin/bitcoin/issues/19303

hodlinator commented on src/net.h:
Could also annotate this one:
```C++
EXCLUSIVE_LOCKS_REQUIRED(!connman.m_nodes_mutex)
```
which leads to also annotating `ThreadMessageHandler`.

hodlinator reviewed (commented):
Reviewed 4aad3714d66cc80c08cb1c827fb49fba7dcf8d50

All annotations make sense, but left Q regarding 1-2 potentially missing ones.

vasild commented on src/net_processing.cpp:
```diff
@@ -1285,23 +1285,23 @@ void PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid)
-                pnodeStop->m_bip152_highbandwidth_to = false;
-                return true;
-            });
-            lNodesAnnouncingHeaderAndIDs.pop_front();
```
Such an `Assume()` would make for stricter requirements, even in `master`, or more complex reasoning to confirm that it does not make stricter requirements. I mean, can `lNodesAnnouncingHeaderAndIDs` have e.g. 5 elements when this code snippet starts? It is not obvious by just looking at this code, so one has to assume that it can or study the code elsewhere to confirm that it cannot have 4 or more.

So, both in `master` and in this PR such an `Assume()` would be triggered if `lNodesAnnouncingHeaderAndIDs` has e.g. 5 elements.

The logic in `master` is - add one object and remove one object if they are too many. This PR keeps it that way.

hodlinator replied:
Agree that one cannot know unless one greps the entire codebase for insertions to `lNodesAnnouncingHeaderAndIDs`. Currently this is the only function.
But worst case, some parallel pull request could introduce another function that inserts more elements than 3. Using a dev-only `Assume` instead of `assert` as I suggested would make it less of a hazard.

Maybe it should be a precondition as well. Anyways, just a minor suggestion to attempt making existing tests against number 3 less magical.

vasild replied:
I considered this again. I like to have the `Assume()` for documentation purposes. But that would make it stricter compared to `master` and this is not the purpose of this PR. I mean - in `master` and in this PR (as it is now) the logic is "if there are more than 3, remove one". Yes, we do imply that there will not be more than 3 at the end, but asserting that (in dev-only `Assume()`) seems to be out of the scope of this PR. So, I will leave it as it is.

hodlinator reviewed (approved):
ACK 4b3a2c23608e709a63b9d5bd69c3eb16cf08e462

Manually went through all functions using `CConnman::m_nodes_mutex` and only remaining issue I could find is now fixed (https://github.com/bitcoin/bitcoin/pull/32394/files#r2433600255).

Also checked out first commit and only cherrypicked mutex type change to re-confirm Clang spews warnings about missing annotations (and no warnings at HEAD of PR).

vasild commented:
`4b3a2c2360...1a9274f391`: rebase due to conflicts

hodlinator reviewed (approved):
re-ACK 1a9274f391df0465c17b6b6664988af70910e39b

Changes since previous review (https://github.com/bitcoin/bitcoin/pull/32394#pullrequestreview-3353773054):
* Resolves conflict with #32983.

Crypt-iQ commented on src/net_processing.cpp:
```diff
@@ -1285,23 +1285,23 @@ void PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid)
+            pnodeStop->m_bip152_highbandwidth_to = false;
+            return true;
+        });
+        lNodesAnnouncingHeaderAndIDs.pop_front();
```
I was initially concerned that maybe `lNodesAnnouncingHeaderAndIDs.front()` could have been removed from in between releasing the lock and acquiring the lock again. It seems fine since if we do not find `lNodesAnnouncingHeaderAndIDs.front()`, we don't need to set `m_bip152_highbandwidth_to` false and we still remove it from the list. There was some previous discussion about the locking [here](https://github.com/bitcoin/bitcoin/pull/22147#issuecomment-855904105).

Crypt-iQ replied:
I think I was a little unclear. I meant that the `CNode*` that `lNodesAnnouncingHeaderAndIDs.front()` refers to might have been removed from in `m_nodes` if we're not holding `m_nodes_mutex` the entire time. Though I did not write a test to check if this is actually possible, so I might be wrong.

Crypt-iQ reviewed (commented):
crACK 1a9274f391df0465c17b6b6664988af70910e39b

Needs a small rebase since #33956 added 2 annotations.

hodlinator reviewed (approved):
re-ACK a6ccd6fdbf838d70ada83b6079112c648c490e1c

Rebase to fix conflict with #33956

sedited commented:
This has been in need of a rebase for a while, but already got a couple of ACKs. Can you push an update?

hodlinator reviewed (approved):
re-ACK e46d0e2c801d4f68624a602d3255ddd392b2dac0

Just changes to diff context and adapting to private broadcast being merged.

---

nit: Might want to also do:
```diff
 
 void CConnman::ThreadPrivateBroadcast()
 {
+    AssertLockNotHeld(m_nodes_mutex);
     AssertLockNotHeld(m_unused_i2p_sessions_mutex);
 
     size_t addrman_num_bad_addresses{0};
```

hodlinator reviewed (approved):
re-ACK 6da753002a7e5749e03ddb7c8879803f3144858e

w0xlt reviewed (commented):
ACK 6da753002a7e5749e03ddb7c8879803f3144858e

Crypt-iQ commented on src/net_processing.cpp:
```diff
@@ -1310,23 +1310,23 @@ void PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid)
         lNodesAnnouncingHeaderAndIDs.push_back(pfrom->GetId());
         return true;
     });
+    if (lNodesAnnouncingHeaderAndIDs.size() > 3) {
```
Yup. Any earlier run of this function will make `lNodesAnnouncingHeaderAndIDs` be at most length 3. Then if the first callback doesn't run because the peer disconnects, it will still be at most 3. Then the second callback won't run.

vasild replied:
That `Assume()` looks unnecessary strong to me. We just want to trim the list if it grew too big. I think in the current code such an `Assume()` would never be triggered, but may be after future changes and there is no really a reason to assert that if the list is too big then we must have had appended the node to it.

Anyway, I changed it to:

```
bool nodeid_was_appended{ForNode(...)};
if (nodeid_was_appended && the list is too big) then trim it
```

sedited reviewed (commented):
Concept ACK

vasild commented:
`6da753002a7e5749e03ddb7c8879803f3144858e...11713c9fa91ca583063750cc826d0138b3a611f6`: address suggestions

> @vasild Could you please address (respond to or implement) the remaining comments here?

Done.

vasild commented on src/net_processing.cpp:
```diff
@@ -1310,23 +1310,23 @@ void PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid)
+    if (lNodesAnnouncingHeaderAndIDs.size() > 3) {
+        // As per BIP152, we only get 3 of our peers to announce
+        // blocks using compact encodings.
+        m_connman.ForNode(lNodesAnnouncingHeaderAndIDs.front(), [this](CNode* pnodeStop){
```
Done

vasild commented on src/net.cpp:
```diff
@@ -3204,6 +3204,7 @@ void CConnman::ThreadI2PAcceptIncoming()
 
 void CConnman::ThreadPrivateBroadcast()
 {
+    AssertLockNotHeld(m_nodes_mutex);
```
Those `AssertLockNotHeld()` were triggered occasionally and I couldn't figure it out, so it was easiest to just remove them.

vasild replied:
Jokes aside, I don't remember a reason for their omission. I have added them now, except for functions whose very first line is `LOCK(m_nodes_mutex);`.

w0xlt reviewed (commented):
ACK 11713c9fa91ca583063750cc826d0138b3a611f6

Crypt-iQ commented:
utACK 11713c9fa91ca583063750cc826d0138b3a611f6

hodlinator commented on src/net.cpp:
```diff
@@ -1736,6 +1739,8 @@ bool CConnman::AttemptToEvictConnection()
 }
 
 void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
```
nanonit: Please consider iff you touch again:
```suggestion
void CConnman::AcceptConnection(const ListenSocket& hListenSocket)
{
```

hodlinator reviewed (approved):
re-ACK 11713c9fa91ca583063750cc826d0138b3a611f6

Changes since previous review (6da753002a7e5749e03ddb7c8879803f3144858e):
* Added `bool nodeid_was_appended` in `PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs` so we only check to see if we should remove an entry if we have added one. I double-checked the code to make sure the removal of entries wasn't also serving some other place where nodes were being added to `lNodesAnnouncingHeaderAndIDs`.
* Added a bunch of `AssertLockNotHeld(m_nodes_mutex)` in the bodies of methods declared with `EXCLUSIVE_LOCKS_REQUIRED(!m_nodes_mutex)`.

sedited reviewed (approved):
ACK 11713c9fa91ca583063750cc826d0138b3a611f6
